Content.Server/GameTicking/Rules/ZombieRuleSystem.cs

```diff
@@ -263,82 +397,169 @@ private List<EntityUid> GetHealthyHumans(bool includeOffStation = true)
-            var inCharacterName = MetaData(ownedEntity).EntityName;
-            var action = new InstantAction(_prototypeManager.Index<InstantActionPrototype>(ZombieRuleComponent.ZombifySelfActionPrototype));
-            _action.AddAction(mind.OwnedEntity.Value, action, null);
-
-            var message = Loc.GetString("zombie-patientzero-role-greeting");
-            var wrappedMessage = Loc.GetString("chat-manager-server-wrap-message", ("message", message));
-
-            //gets the names now in case the players leave.
-            //this gets unhappy if people with the same name get chosen. Probably shouldn't happen.
-            component.InitialInfectedNames.Add(inCharacterName, zombie.Name);
-
-            // I went all the way to ChatManager.cs and all i got was this lousy T-shirt
-            // You got a free T-shirt!?!?
-            _chatManager.ChatMessageToOne(Shared.Chat.ChatChannel.Server, message,
-               wrappedMessage, default, false, zombie.ConnectedClient, Color.Plum);
-            _audio.PlayGlobal(component.InitialInfectedSound, ownedEntity);
+
+            if (mind.Session != null)
+            {
+                var message = Loc.GetString("zombie-patientzero-role-greeting");
+                var wrappedMessage = Loc.GetString("chat-manager-server-wrap-message", ("message", message));
+
+                //gets the names now in case the players leave.
+                rules.InitialInfectedNames[inCharacterName] = mind.Session.Name;
+
+                // I went all the way to ChatManager.cs and all i got was this lousy T-shirt
+                // You got a free T-shirt!?!?
+                _chatManager.ChatMessageToOne(Shared.Chat.ChatChannel.Server, message,
+                   wrappedMessage, default, false, mind.Session.ConnectedClient, Color.Plum);
+
+                // Notify player about new role assignment with a sound effect
+                _audioSystem.PlayGlobal(rules.InitialInfectedSound, mind.Session);
+            }
         }
     }
+
+    public (EntityUid, ZombieRuleComponent?) FindActiveRule()
+    {
+        var query = EntityQueryEnumerator<ZombieRuleComponent, GameRuleComponent>();
+
+        while (query.MoveNext(out var uid, out var zombies, out var gameRule))
+        {
+            if (!GameTicker.IsGameRuleActive(uid, gameRule))
+                continue;
+
+            return (uid, zombies);
+        }
+
+        return (EntityUid.Invalid, null);
+    }
+
+    public void InfectPlayer(EntityUid uid)
+    {
+        var zombie = EnsureComp<ZombieComponent>(uid);
+        var (rulesUid, rules) = FindActiveRule();
+
+        if (rules != null)
+        {
+            // Admin command added a new zombie to this existing rule.
+            zombie.Settings = rules.EarlySettings;
+            zombie.VictimSettings = rules.VictimSettings;
+            zombie.Family = new ZombieFamily(){ Rules = rulesUid, Generation = 0 };
+        }
+
+        _initialZombie.ForceInfection(uid, zombie);
+    }
```

Comment:
Doesn't this defeat the point of having game rules as entities?

Reply:
I'm not sure how? It just makes it easier to track which rule created which zombies. That lets us handle all sorts of edge cases around how many initial infected if multiple zombie events run.

With this PR (assuming I don't strip out ZombieFamily) each event gets its own initial infected and tracks infections seperately. Different events could also have different zombie virulence. Admin zombification victims get added to the running event for better end round detection / end screen visuals.

Reply:
FindActiveRule only considers 1 zombie gamerule ever existing but it needs to be able to support as many as we want being added.